syntax = "proto3";

package silenteight.agent.allowlist.v1.api;

import "google/api/field_behavior.proto";

option java_package = "com.silenteight.proto.agent.allowlist.v1.api";
option java_outer_classname = "AllowListAgentApiProto";
option java_multiple_files = true;

service AllowListAgent {

  rpc Verify (VerifyRequest) returns (VerifyResponse) {
  }
}

message VerifyRequest {
  repeated string characteristics_value = 1 [(google.api.field_behavior) = REQUIRED];
  repeated string allow_list_name = 2 [(google.api.field_behavior) = REQUIRED];
}

message VerifyResponse {
  string solution = 1 [(google.api.field_behavior) = REQUIRED];
  AllowListReason reason = 2 [(google.api.field_behavior) = REQUIRED];
}

message AllowListReason {
  repeated MatchedCharacteristic matched_characteristics = 1 [(google.api.field_behavior) = REQUIRED];
  Configuration configuration = 2 [(google.api.field_behavior) = REQUIRED];
}

message MatchedCharacteristic {
  string characteristic_value = 1 [(google.api.field_behavior) = REQUIRED];
  string solution = 2 [(google.api.field_behavior) = REQUIRED];
  uint32 false_positive_count = 3 [(google.api.field_behavior) = REQUIRED];
  uint32 true_positive_count = 4 [(google.api.field_behavior) = REQUIRED];
  // false positive by true positive ratio
  double ratio = 5 [(google.api.field_behavior) = REQUIRED];
}

message Configuration {
  string allow_list_name = 1 [(google.api.field_behavior) = REQUIRED];
  uint32 min_false_positive_count = 2 [(google.api.field_behavior) = REQUIRED];
  double min_ratio = 3 [(google.api.field_behavior) = REQUIRED];
}
